File: leetcode/动态规划/0070.py
# 70. 爬楼梯
# https://leetcode-cn.com/problems/climbing-stairs/

# 3. 动态规划 滚动数组 (只保留3个变量)
# 执行用时: 16 ms , 在所有 Python 提交中击败了 86.14% 的用户
# 内存消耗: 12.9 MB , 在所有 Python 提交中击败了 6.67% 的用户
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        p = 0
        q = 0
        r = 1
        for i in range(1, n + 1):
            p = q
            q = r
            r = p + q
        return r


# 2. 动态规划 一维数组
# 执行用时: 20 ms , 在所有 Python 提交中击败了 66.55% 的用户
# 内存消耗: 12.7 MB , 在所有 Python 提交中击败了 6.67% 的用户
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [1, 1]
        for i in range(2, n + 1):
            dp.append(dp[i - 1] + dp[i - 2])
        return dp[n]


# 1. 递归
# 执行结果: 超时
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        each_max = 2
        if n == 0:
            return 1
        if n < 0:
            return 0
        sum = 0
        for i in range(each_max):
            each = i + 1
            sum = sum + self.climbStairs(n - each)
        return sum

File: leetcode/动态规划/0070.java

// 爬楼梯
// https://leetcode-cn.com/problems/climbing-stairs/

// 2. 动态规划 滚动数组 (只保留3个变量)
class Solution {
  public int climbStairs(int n) {
    // ...
  }
}

// 1. 动态规划 一维数组
// 执行用时 : 0 ms , 在所有 Java 提交中击败了 100.00% 的用户
// 内存消耗 : 36.1 MB , 在所有 Java 提交中击败了 5.66% 的用户
class Solution {
  public int climbStairs(int n) {
    int[] dp = new int[n + 1];
    dp[0] = 1;
    dp[1] = 1;
    for (int i = 2; i <= n; i++)
      dp[i] = dp[i - 1] + dp[i - 2];
    return dp[n];
  }
}
